Improved performance of the safe prime generator

pdyraga opened the pull request:
Refs https://github.com/keep-network/keep-core/issues/115

In order to generate a threshold Paillier key, we need to generate two safe primes `p` and `q`. For our use cases, `p` and `q` must be at least-1024 bit numbers.

This PR aims at improving performance of the safe prime generator. The previous approach was as follows:
```
1. Generate prime `q` of `nBits - 1`
2. Compute `p = 2q + 1`.
3. Check if `p` is prime. If it is, return `(p,q)`. If not, go back to the point 1.
```

This approach was simple, but not fast enough. In this PR, we propose an alternative algorithm as well as make the search process concurrent.

The new algorithm is as follows:
```
1. Generate a random odd number `q` of length `pBitLen-1` with two the most
   significant bytes set to `1`.
2. Execute preliminary primality test on `q` checking whether it is coprime
   to all the elements of `smallPrimes`. It allows to eliminate trivial
   cases quickly, when `q` is obviously no prime, without running an
   expensive final primality tests.
   If `q` is coprime to all of the `smallPrimes`, then go to the point 3.
   If not, add `2` and try again. Do it at most 10 times.
3. Check the potentially prime `q`, whether `q = 1 (mod 3)`. This will
   happen for 50% of cases.
   If it is, then `p = 2q+1` will be a multiple of 3, so it will be obviously
   not a prime number. In this case, add `2` and try again. Do it at most 10
   times. If `q != 1 (mod 3)`, go to the point 4.
4. Now we know `q` is potentially prime and `p = 2q+1` is not a multiple of
   3. We execute a preliminary primality test on `p`, checking whether
   it is coprime to all the elements of `smallPrimes` just like we did for
   `q` in point 2. If `p` is not coprime to at least one element of the
   `smallPrimes`, then go back to point 1.
   If `p` is coprime to all the elements of `smallPrimes`, go to point 5.
5. At this point, we know `q` is potentially prime, and `p=q+1` is also
   potentially prime. We need to execute a final primality test for `q`.
   We apply Miller-Rabin and Baillie-PSW tests. If they succeed, it means
   that `q` is prime with a very high probability. Knowing `q` is prime,
   we use Pocklington's criterion to prove the primality of `p=2q+1`, that
   is, we execute Fermat primality test to base 2 checking whether
   `2^{p-1} = 1 (mod p)`. It's significantly faster than running full
   Miller-Rabin and Baillie-PSW for `p`.
   If `q` and `p` are found to be prime, return them as a result. If not, go
   back to the point 1.
```

How fast we generate a prime number is mostly a matter of luck and it depends on how lucky we are with choosing the first bytes. With today's multicore processors, we can execute the process on multiple cores concurrently, accept the first valid result and cancel the rest of work. This way, with the same finding algorithm, we can get the result faster. Such an approach is presented in this PR.

Detailed benchmark results are presented in the comment. We can observe an improvement in almost all cases (still, a small factor of luck is needed). I've tested for `512`, `1024` and `2048` bit numbers on my MBP 3,1 GHz Intel Core i7 with concurrency level set to 4. 
I did not present a result for `2048` bits, since the previous algorithm wasn't able to produce one `2048`bit safe prime in one hour so I had nothing to compare.

If you want, you can compare the new approach with the performance of `openssl dhparam -5 2048`. (512/1024).

pdyraga committed:
Draft implementation of a safe prime generator

I've adapted a Go code for generating random numbers by inserting some
optimisations that will allow us to generate safe primes faster than
with the previous, naive approach.

First of all, having q which can be prime, we first check whether q%3=1.
If that's true, there is no chance p=2q+1 is prime. It lets us to reject
candidate numbers quicker without running an expensive primality tests.

Also, before we run a primality test for q, we may check p=2q+1 against
the primes between 3-53 (We are limited by Go's uint64 range).

If all those conditions are met and we know p is prime, it's enough to
check Pocklington criterion for q instead of running an expensive
primality test for it.

pdyraga committed:
Generate safe prime concurrently

How fast we generate a prime number is mostly a matter of luck. It
varies from miliseconds to seconds and depends how lucky we are with
drawing bytes.

With today's multicore processors, we can execute the process on
multiple cores concurrently, accept the first valid result and cancel
the rest of work. This way, with the same algorithm, we get the result
faster.

pdyraga committed:
Improved documentation for the safe prime generator

pdyraga committed:
Wait for all goroutines to finish

Initially, we did not wait for all goroutines to finish which was fine
as long as we did not run all tests with `go test` and observe deadlock
in 50% of cases. This is interesting, since deadlock never occurred when
running safe_prime_generator_test.go. Anyway, waiting for all go
routines to finish shouldn't harm the performance a lot and it gives a
guarantee we did not miss anything.

pdyraga committed:
Use the new, faster way for generating safe primes everywhere

pdyraga commented:
# 512-bit prime

```
benchmark               old ns/op       new ns/op     delta
BenchmarkPrimeGen-8     2268969681      188363314     -91.70%
BenchmarkPrimeGen-8     11386138267     170493155     -98.50%
BenchmarkPrimeGen-8     2631451953      180446766     -93.14%
BenchmarkPrimeGen-8     1934699051      208869584     -89.20%
BenchmarkPrimeGen-8     10074871723     305999350     -96.96%
BenchmarkPrimeGen-8     5705400096      260922273     -95.43%
BenchmarkPrimeGen-8     5569562094      153497360     -97.24%
BenchmarkPrimeGen-8     1784282054      210098300     -88.23%
BenchmarkPrimeGen-8     1239902622      209132014     -83.13%
BenchmarkPrimeGen-8     2478345181      216459976     -91.27%
BenchmarkPrimeGen-8     3105669082      186708939     -93.99%
BenchmarkPrimeGen-8     2015272750      168540463     -91.64%
BenchmarkPrimeGen-8     2052936393      250378808     -87.80%
BenchmarkPrimeGen-8     2870424928      325760690     -88.65%
BenchmarkPrimeGen-8     6326295939      203200026     -96.79%
BenchmarkPrimeGen-8     6128458193      257232450     -95.80%
BenchmarkPrimeGen-8     1991237139      178546016     -91.03%
BenchmarkPrimeGen-8     4419054637      215965061     -95.11%
BenchmarkPrimeGen-8     1278097111      150693571     -88.21%
BenchmarkPrimeGen-8     1596391885      186586729     -88.31%
BenchmarkPrimeGen-8     2874842860      210377943     -92.68%
BenchmarkPrimeGen-8     2763618377      205416284     -92.57%
BenchmarkPrimeGen-8     1894357212      234921937     -87.60%
BenchmarkPrimeGen-8     2820765637      313129768     -88.90%
BenchmarkPrimeGen-8     5442844899      213699420     -96.07%
BenchmarkPrimeGen-8     2317770026      327961987     -85.85%
BenchmarkPrimeGen-8     13632503186     254280747     -98.13%
BenchmarkPrimeGen-8     1627820128      149842621     -90.79%
BenchmarkPrimeGen-8     1618299835      233881054     -85.55%
BenchmarkPrimeGen-8     1563058438      244266049     -84.37%
BenchmarkPrimeGen-8     6981079384      153383777     -97.80%
BenchmarkPrimeGen-8     8583457685      232863218     -97.29%
BenchmarkPrimeGen-8     2920009879      182719931     -93.74%
BenchmarkPrimeGen-8     2855556611      316134549     -88.93%
BenchmarkPrimeGen-8     4192479325      239415232     -94.29%
BenchmarkPrimeGen-8     2124136761      270599184     -87.26%
BenchmarkPrimeGen-8     3354065514      226163593     -93.26%
BenchmarkPrimeGen-8     4793574227      195853126     -95.91%
BenchmarkPrimeGen-8     7926695778      203830914     -97.43%
BenchmarkPrimeGen-8     3783685714      297339207     -92.14%
BenchmarkPrimeGen-8     5769461617      193390742     -96.65%
BenchmarkPrimeGen-8     3899060159      175458617     -95.50%
BenchmarkPrimeGen-8     8162547820      395694820     -95.15%
BenchmarkPrimeGen-8     1162491142      210804576     -81.87%
BenchmarkPrimeGen-8     1292424634      276726318     -78.59%
BenchmarkPrimeGen-8     2615398761      184457813     -92.95%
BenchmarkPrimeGen-8     3015784825      249611044     -91.72%
BenchmarkPrimeGen-8     2652024361      265722298     -89.98%
BenchmarkPrimeGen-8     2649694963      496443957     -81.26%
BenchmarkPrimeGen-8     1152981067      249966541     -78.32%

benchmark               old allocs     new allocs     delta
BenchmarkPrimeGen-8     362387         377422         +4.15%
BenchmarkPrimeGen-8     1769941        349798         -80.24%
BenchmarkPrimeGen-8     419616         373273         -11.04%
BenchmarkPrimeGen-8     311689         415817         +33.41%
BenchmarkPrimeGen-8     1653901        585598         -64.59%
BenchmarkPrimeGen-8     947327         502208         -46.99%
BenchmarkPrimeGen-8     917727         288982         -68.51%
BenchmarkPrimeGen-8     293204         398834         +36.03%
BenchmarkPrimeGen-8     204671         394669         +92.83%
BenchmarkPrimeGen-8     412135         399917         -2.96%
BenchmarkPrimeGen-8     512568         325195         -36.56%
BenchmarkPrimeGen-8     336663         282991         -15.94%
BenchmarkPrimeGen-8     339021         427862         +26.21%
BenchmarkPrimeGen-8     475326         551846         +16.10%
BenchmarkPrimeGen-8     1045261        355640         -65.98%
BenchmarkPrimeGen-8     1009762        438359         -56.59%
BenchmarkPrimeGen-8     328106         306610         -6.55%
BenchmarkPrimeGen-8     727409         368051         -49.40%
BenchmarkPrimeGen-8     202614         258985         +27.82%
BenchmarkPrimeGen-8     263268         313234         +18.98%
BenchmarkPrimeGen-8     470412         355386         -24.45%
BenchmarkPrimeGen-8     444871         339915         -23.59%
BenchmarkPrimeGen-8     325247         391494         +20.37%
BenchmarkPrimeGen-8     496687         491867         -0.97%
BenchmarkPrimeGen-8     932339         340511         -63.48%
BenchmarkPrimeGen-8     382644         521179         +36.20%
BenchmarkPrimeGen-8     2351353        356008         -84.86%
BenchmarkPrimeGen-8     281696         239016         -15.15%
BenchmarkPrimeGen-8     284146         384130         +35.19%
BenchmarkPrimeGen-8     271630         382155         +40.69%
BenchmarkPrimeGen-8     1205117        199428         -83.45%
BenchmarkPrimeGen-8     1510913        368338         -75.62%
BenchmarkPrimeGen-8     501141         305767         -38.99%
BenchmarkPrimeGen-8     497295         514941         +3.55%
BenchmarkPrimeGen-8     744680         363065         -51.25%
BenchmarkPrimeGen-8     370638         391459         +5.62%
BenchmarkPrimeGen-8     601074         344112         -42.75%
BenchmarkPrimeGen-8     846588         294090         -65.26%
BenchmarkPrimeGen-8     1353487        309463         -77.14%
BenchmarkPrimeGen-8     638008         479816         -24.79%
BenchmarkPrimeGen-8     1012358        318190         -68.57%
BenchmarkPrimeGen-8     688078         281367         -59.11%
BenchmarkPrimeGen-8     1424913        595065         -58.24%
BenchmarkPrimeGen-8     200156         333529         +66.63%
BenchmarkPrimeGen-8     224154         365020         +62.84%
BenchmarkPrimeGen-8     445941         279446         -37.34%
BenchmarkPrimeGen-8     535468         377860         -29.43%
BenchmarkPrimeGen-8     458854         386878         -15.69%
BenchmarkPrimeGen-8     459767         769193         +67.30%
BenchmarkPrimeGen-8     200718         373180         +85.92%

benchmark               old bytes     new bytes     delta
BenchmarkPrimeGen-8     82728176      37112556      -55.14%
BenchmarkPrimeGen-8     409631096     34535683      -91.57%
BenchmarkPrimeGen-8     96872624      36726974      -62.09%
BenchmarkPrimeGen-8     71888744      41056451      -42.89%
BenchmarkPrimeGen-8     378938424     57864424      -84.73%
BenchmarkPrimeGen-8     217741152     49683195      -77.18%
BenchmarkPrimeGen-8     212017904     28595818      -86.51%
BenchmarkPrimeGen-8     67207840      39318799      -41.50%
BenchmarkPrimeGen-8     47274152      38828333      -17.87%
BenchmarkPrimeGen-8     95460120      39496033      -58.63%
BenchmarkPrimeGen-8     118061851     32182768      -72.74%
BenchmarkPrimeGen-8     78798712      27910800      -64.58%
BenchmarkPrimeGen-8     78144912      42148965      -46.06%
BenchmarkPrimeGen-8     108414480     54402728      -49.82%
BenchmarkPrimeGen-8     241962472     35135193      -85.48%
BenchmarkPrimeGen-8     232679504     43332865      -81.38%
BenchmarkPrimeGen-8     74858416      30346100      -59.46%
BenchmarkPrimeGen-8     167936024     36391868      -78.33%
BenchmarkPrimeGen-8     46505936      25568655      -45.02%
BenchmarkPrimeGen-8     60196144      30969552      -48.55%
BenchmarkPrimeGen-8     108738512     35072412      -67.75%
BenchmarkPrimeGen-8     102350984     33572493      -67.20%
BenchmarkPrimeGen-8     74867877      38710753      -48.29%
BenchmarkPrimeGen-8     114624815     48575358      -57.62%
BenchmarkPrimeGen-8     218440752     33697121      -84.57%
BenchmarkPrimeGen-8     86905816      51420408      -40.83%
BenchmarkPrimeGen-8     541723888     35038903      -93.53%
BenchmarkPrimeGen-8     64802296      23532448      -63.69%
BenchmarkPrimeGen-8     64726184      37951417      -41.37%
BenchmarkPrimeGen-8     62328698      37739038      -39.45%
BenchmarkPrimeGen-8     279747264     19709518      -92.95%
BenchmarkPrimeGen-8     347825616     36348116      -89.55%
BenchmarkPrimeGen-8     114803440     30225522      -73.67%
BenchmarkPrimeGen-8     113669288     50717204      -55.38%
BenchmarkPrimeGen-8     172993448     35961875      -79.21%
BenchmarkPrimeGen-8     84498152      38555288      -54.37%
BenchmarkPrimeGen-8     138581872     34002839      -75.46%
BenchmarkPrimeGen-8     193495360     29064355      -84.98%
BenchmarkPrimeGen-8     311015920     30525862      -90.19%
BenchmarkPrimeGen-8     147068656     47471092      -67.72%
BenchmarkPrimeGen-8     234513488     31368265      -86.62%
BenchmarkPrimeGen-8     159871800     27874576      -82.56%
BenchmarkPrimeGen-8     329108176     58733391      -82.15%
BenchmarkPrimeGen-8     45833152      32946302      -28.12%
BenchmarkPrimeGen-8     51570752      36111451      -29.98%
BenchmarkPrimeGen-8     102344696     27662710      -72.97%
BenchmarkPrimeGen-8     123446618     37352648      -69.74%
BenchmarkPrimeGen-8     105297912     38243751      -63.68%
BenchmarkPrimeGen-8     106156028     75890648      -28.51%
BenchmarkPrimeGen-8     45458288      36981435      -18.65%
```

pdyraga commented:
# 1024-bit prime

```
benchmark               old ns/op        new ns/op       delta
BenchmarkPrimeGen-8     36861702183      5566762010      -84.90%
BenchmarkPrimeGen-8     45959705486      2908156086      -93.67%
BenchmarkPrimeGen-8     32302799509      1600852073      -95.04%
BenchmarkPrimeGen-8     10559412608      6909158605      -34.57%
BenchmarkPrimeGen-8     26361757373      2518036679      -90.45%
BenchmarkPrimeGen-8     29501857227      1448072197      -95.09%
BenchmarkPrimeGen-8     231272508288     1392535303      -99.40%
BenchmarkPrimeGen-8     7657751078       10036763187     +31.07%
BenchmarkPrimeGen-8     51558997401      8023889630      -84.44%
BenchmarkPrimeGen-8     16745757300      2572674783      -84.64%
BenchmarkPrimeGen-8     28182579779      2706935779      -90.40%
BenchmarkPrimeGen-8     19612730893      1771794157      -90.97%
BenchmarkPrimeGen-8     24840617742      1559935715      -93.72%
BenchmarkPrimeGen-8     2466813963       4869175794      +97.39%
BenchmarkPrimeGen-8     8494698198       5458197106      -35.75%
BenchmarkPrimeGen-8     24406594314      8387214807      -65.64%
BenchmarkPrimeGen-8     111450059549     2289543448      -97.95%
BenchmarkPrimeGen-8     9711480641       1756765315      -81.91%
BenchmarkPrimeGen-8     81936301479      2577900911      -96.85%
BenchmarkPrimeGen-8     2767979308       3723179272      +34.51%
BenchmarkPrimeGen-8     101696370631     3567358343      -96.49%
BenchmarkPrimeGen-8     22492110644      2786609398      -87.61%
BenchmarkPrimeGen-8     84566338764      871273575       -98.97%
BenchmarkPrimeGen-8     17797979625      10846582551     -39.06%
BenchmarkPrimeGen-8     28036722072      1107767565      -96.05%
BenchmarkPrimeGen-8     26167954580      3447014075      -86.83%
BenchmarkPrimeGen-8     16405363943      1506522554      -90.82%
BenchmarkPrimeGen-8     4533869408       3009052149      -33.63%
BenchmarkPrimeGen-8     28292850899      2939190707      -89.61%
BenchmarkPrimeGen-8     58376102081      7330703326      -87.44%
BenchmarkPrimeGen-8     137515732119     2121113847      -98.46%
BenchmarkPrimeGen-8     5343562821       1170653779      -78.09%
BenchmarkPrimeGen-8     4214167930       5996486300      +42.29%
BenchmarkPrimeGen-8     57539380856      4818339856      -91.63%
BenchmarkPrimeGen-8     56463189239      7023259834      -87.56%
BenchmarkPrimeGen-8     31148876342      2628034774      -91.56%
BenchmarkPrimeGen-8     35924957141      10139439977     -71.78%
BenchmarkPrimeGen-8     37037673636      5741866015      -84.50%
BenchmarkPrimeGen-8     123071912422     15779098285     -87.18%
BenchmarkPrimeGen-8     18735582572      1191203409      -93.64%
BenchmarkPrimeGen-8     14950810812      3990970414      -73.31%
BenchmarkPrimeGen-8     45614828915      5893157572      -87.08%
BenchmarkPrimeGen-8     38343598558      2582931037      -93.26%
BenchmarkPrimeGen-8     37089393514      24885103035     -32.91%
BenchmarkPrimeGen-8     21408341771      1005562686      -95.30%
BenchmarkPrimeGen-8     82520362962      12180098905     -85.24%
BenchmarkPrimeGen-8     71017278949      5728216833      -91.93%
BenchmarkPrimeGen-8     34683384212      10680677166     -69.21%
BenchmarkPrimeGen-8     176688401313     1472591237      -99.17%
BenchmarkPrimeGen-8     27576279959      2510611582      -90.90%

benchmark               old allocs     new allocs     delta
BenchmarkPrimeGen-8     1296874        2776508        +114.09%
BenchmarkPrimeGen-8     1629910        1384754        -15.04%
BenchmarkPrimeGen-8     1145560        816307         -28.74%
BenchmarkPrimeGen-8     375806         3260984        +767.73%
BenchmarkPrimeGen-8     936175         1139303        +21.70%
BenchmarkPrimeGen-8     1032492        677510         -34.38%
BenchmarkPrimeGen-8     7860859        664895         -91.54%
BenchmarkPrimeGen-8     278511         4575811        +1542.96%
BenchmarkPrimeGen-8     1772920        3768429        +112.55%
BenchmarkPrimeGen-8     599647         1210124        +101.81%
BenchmarkPrimeGen-8     974381         1260570        +29.37%
BenchmarkPrimeGen-8     684946         796530         +16.29%
BenchmarkPrimeGen-8     895839         692513         -22.70%
BenchmarkPrimeGen-8     84866          2213842        +2508.63%
BenchmarkPrimeGen-8     292856         2442000        +733.86%
BenchmarkPrimeGen-8     865260         3841449        +343.96%
BenchmarkPrimeGen-8     3960309        972024         -75.46%
BenchmarkPrimeGen-8     348814         782942         +124.46%
BenchmarkPrimeGen-8     2872539        1177839        -59.00%
BenchmarkPrimeGen-8     96468          1611606        +1570.61%
BenchmarkPrimeGen-8     3535318        1598072        -54.80%
BenchmarkPrimeGen-8     774395         1221268        +57.71%
BenchmarkPrimeGen-8     2807298        369722         -86.83%
BenchmarkPrimeGen-8     551587         4623520        +738.22%
BenchmarkPrimeGen-8     909261         477328         -47.50%
BenchmarkPrimeGen-8     859315         1338424        +55.75%
BenchmarkPrimeGen-8     549266         617694         +12.46%
BenchmarkPrimeGen-8     155427         1276826        +721.50%
BenchmarkPrimeGen-8     976922         1315785        +34.69%
BenchmarkPrimeGen-8     2063480        3193686        +54.77%
BenchmarkPrimeGen-8     4841641        935303         -80.68%
BenchmarkPrimeGen-8     183380         517693         +182.31%
BenchmarkPrimeGen-8     147059         2375516        +1515.35%
BenchmarkPrimeGen-8     2041437        2040157        -0.06%
BenchmarkPrimeGen-8     1975536        2874974        +45.53%
BenchmarkPrimeGen-8     1074994        1082881        +0.73%
BenchmarkPrimeGen-8     1287402        3817244        +196.51%
BenchmarkPrimeGen-8     1341691        2100462        +56.55%
BenchmarkPrimeGen-8     4409132        6425913        +45.74%
BenchmarkPrimeGen-8     658803         463552         -29.64%
BenchmarkPrimeGen-8     516970         1388957        +168.67%
BenchmarkPrimeGen-8     1628919        2127240        +30.59%
BenchmarkPrimeGen-8     1358365        980612         -27.81%
BenchmarkPrimeGen-8     1300251        9751724        +649.99%
BenchmarkPrimeGen-8     773588         374964         -51.53%
BenchmarkPrimeGen-8     2933710        5006827        +70.67%
BenchmarkPrimeGen-8     2551294        2293989        -10.09%
BenchmarkPrimeGen-8     1139499        4228107        +271.05%
BenchmarkPrimeGen-8     6096582        549261         -90.99%
BenchmarkPrimeGen-8     974002         1009220        +3.62%

benchmark               old bytes      new bytes      delta
BenchmarkPrimeGen-8     410531984      344987328      -15.97%
BenchmarkPrimeGen-8     516871536      172012104      -66.72%
BenchmarkPrimeGen-8     363488520      100463880      -72.36%
BenchmarkPrimeGen-8     118047720      404344976      +242.53%
BenchmarkPrimeGen-8     297614840      141266872      -52.53%
BenchmarkPrimeGen-8     328461424      83951088       -74.44%
BenchmarkPrimeGen-8     2498362152     83168312       -96.67%
BenchmarkPrimeGen-8     88981416       568617328      +539.03%
BenchmarkPrimeGen-8     562466976      470503328      -16.35%
BenchmarkPrimeGen-8     190238184      149378720      -21.48%
BenchmarkPrimeGen-8     309543552      156230640      -49.53%
BenchmarkPrimeGen-8     217364304      98794060       -54.55%
BenchmarkPrimeGen-8     284504936      86219460       -69.69%
BenchmarkPrimeGen-8     26543528       274414312      +933.83%
BenchmarkPrimeGen-8     92278832       304842736      +230.35%
BenchmarkPrimeGen-8     273084064      476264192      +74.40%
BenchmarkPrimeGen-8     1260919112     120177016      -90.47%
BenchmarkPrimeGen-8     110596312      97894424       -11.48%
BenchmarkPrimeGen-8     908471736      146710656      -83.85%
BenchmarkPrimeGen-8     31164144       201423000      +546.33%
BenchmarkPrimeGen-8     1126090384     197493728      -82.46%
BenchmarkPrimeGen-8     245900104      152274684      -38.07%
BenchmarkPrimeGen-8     892851912      45775300       -94.87%
BenchmarkPrimeGen-8     176047536      571741624      +224.77%
BenchmarkPrimeGen-8     289382520      59191272       -79.55%
BenchmarkPrimeGen-8     273224016      168431424      -38.35%
BenchmarkPrimeGen-8     173397688      77421576       -55.35%
BenchmarkPrimeGen-8     48816216       159746704      +227.24%
BenchmarkPrimeGen-8     309511600      162808000      -47.40%
BenchmarkPrimeGen-8     654125952      396621240      -39.37%
BenchmarkPrimeGen-8     1536756272     116150672      -92.44%
BenchmarkPrimeGen-8     57621800       63679392       +10.51%
BenchmarkPrimeGen-8     46709232       295358932      +532.34%
BenchmarkPrimeGen-8     646374992      252268344      -60.97%
BenchmarkPrimeGen-8     624530576      357305168      -42.79%
BenchmarkPrimeGen-8     342230304      135394928      -60.44%
BenchmarkPrimeGen-8     411465472      475151664      +15.48%
BenchmarkPrimeGen-8     427305360      262550304      -38.56%
BenchmarkPrimeGen-8     1403440376     797594248      -43.17%
BenchmarkPrimeGen-8     208987312      56730160       -72.85%
BenchmarkPrimeGen-8     163475552      172827720      +5.72%
BenchmarkPrimeGen-8     516464392      263352608      -49.01%
BenchmarkPrimeGen-8     432335688      122748888      -71.61%
BenchmarkPrimeGen-8     414238240      1212255000     +192.65%
BenchmarkPrimeGen-8     246773984      47038352       -80.94%
BenchmarkPrimeGen-8     927995536      623209176      -32.84%
BenchmarkPrimeGen-8     813041256      284473552      -65.01%
BenchmarkPrimeGen-8     360886104      523694720      +45.11%
BenchmarkPrimeGen-8     1934911896     69402096       -96.41%
BenchmarkPrimeGen-8     309722576      125312600      -59.54%
```

pdyraga commented:
The build is failing because Go `1.7` [moved](https://golang.org/doc/go1.7#context) the `golang.org/x/net/context` package into the standard library as `context` and the build configuration runs on Go `1.3`, `1.4` and `tip`.

I don't see any simple solution for having a backward compatibility and not sacrificing quality. Since Go 1.7 was released 2 years ago, I think we can drop the support for `1.3` and `1.4`. 

We'll need to talk with author of the upstream if he agrees on that and we can merge our changes back or if we cut from the upstream and maintain our own version from now on.

pdyraga committed:
Drop support for Go 1.3, 1,4. Support Go 1.7+

Go 1.7 moved the `golang.org/x/net/context` package into the standard
library as `context`. Code importing `context` is no longer compiling
for Go versions earlier than 1.7. Since 1.7 was released two years ago
and we don't want to sacrifice quality, we decided to drop support for
1.4 and 1.3 and support 1.7+ now.

pdyraga commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+	if bitLen < 6 {
+		return nil, nil, errors.New("safe prime size must be at least 6 bits")
+	}
+
```
@keep-network/go I'd appreciate if some more Go experienced folks could take a look at the way goroutines communicate, how channels are created and closed, how the wait group is used here etc.

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+// The code is the original Go implementation of rand.Prime optimized for
+// generating safe (Sophie Germain) primes.
+// A safe prime is a prime number of the form 2p + 1, where p is also a prime.
+
```
Do we need this newline?

pdyraga replied:
Yeah, otherwise it complains 
> package comment should be of the form "Package paillier ..."

I don't want to comment what the `paillier` package is for. I want to give a brief information on what was changed comparing to Go implementation.

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+//
+// The algorithm is as follows:
+// 1. Generate a random odd number `q` of length `pBitLen-1` with two the most
+//    significant bytes set to `1`.
```
Shouldn't it be bits?

pdyraga replied:
Yes, it should!

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+//    If `p` is coprime to all the elements of `smallPrimes`, go to point 5.
+// 5. At this point, we know `q` is potentially prime, and `p=q+1` is also
+//    potentially prime. We need to execute a final primality test for `q`.
+//    We apply Miller-Rabin and Baillie-PSW tests. If they succeeds, it means
```
Shouldn't it be `If they succeed ...`?

pdyraga replied:
Yep!

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+	pBitLen int,
+) {
+	qBitLen := pBitLen - 1
+	b := uint(qBitLen % 8)
```
Maybe explain what `b` is?

pdyraga replied:
I don't want to say something stupid here since it's from the original rand.crypto implementation, but... I think it's used for byte alignment if the expected prime bit length is not a multiple of 8.

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+						q.Add(q, bigMod)
+					}
+
+					qMod3 := new(big.Int).Mod(q, big.NewInt(3))
```
I know it's documented above the function but maybe we could also add a comment here for this qMod3 calculation?

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+
+				q.SetBytes(bytes)
+
+				p.Mul(q, big.NewInt(2))
```
Do we need it here? We have the same in L243

pdyraga replied:
Yeah, good catch. The one in L243 is needed becase we need to re-evaluate `p` before we execute preliminary primality test on it. The one in L213  not really needed. It's probably some leftover of one of my experiments with this code.

nkuba commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,293 @@
+func isPrimeCandidate(number *big.Int) bool {
+	bigMod := new(big.Int)
+	bigMod.Mod(number, smallPrimesProduct)
+	m := bigMod.Uint64()
```
Would doing it in one line without `bigMod` be any better? 
`m := new(big.Int).Mod(number, smallPrimesProduct).Uint64()`

nkuba commented on safe_prime_generator_test.go:
```diff
@@ -0,0 +1,98 @@
+				}
+
+				if len(p.Bytes()) != test.bitLen {
+					fmt.Errorf(
```
Should be `t.Errorf`

pdyraga replied:
![6gehjsv](https://user-images.githubusercontent.com/4712360/43327957-1f80e29e-91bd-11e8-89a9-f730d74c4b9e.gif)

nkuba commented on safe_prime_generator_test.go:
```diff
@@ -0,0 +1,98 @@
+				}
+
+				if len(q.Bytes()) != test.bitLen-1 {
+					fmt.Errorf(
```
Should be `t.Errorf`

nkuba commented on safe_prime_generator_test.go:
```diff
@@ -0,0 +1,98 @@
+					t.Errorf("2q + 1 != p")
+				}
+
+				if len(p.Bytes()) != test.bitLen {
```
It ain't gonna work

nkuba commented on safe_prime_generator_test.go:
```diff
@@ -0,0 +1,98 @@
+					)
+				}
+
+				if len(q.Bytes()) != test.bitLen-1 {
```
Same as above

nkuba commented on utils.go:
```diff
@@ -22,18 +23,11 @@ func Factorial(n int) *big.Int {
-			return
-		}
-	}
+	return GenerateSafePrime(nbits, 4, 120*time.Second)
```
I'm not sure about this `GenerateSafePrimes` function to call `GenerateSafePrime`, it's little bit duplication. Do we really need to keep this `GenerateSafePrimes` function instead of switching our tecdsa implementation to use the new function?

pdyraga replied:
Yeah, good point. It's not JVM with people locked on shapshot, we are referring to a specific version anyway. No need to be such super backward-compatible. Especially that we want to drop a support for 1.3 / 1.4.

nkuba commented on utils.go:
```diff
@@ -22,18 +23,11 @@ func Factorial(n int) *big.Int {
+//
+// Deprecated: This function has been left here just for backward compatibility.
+// Please use `GenerateSafePrime` from the `safe_prime_generator.go` directly.
 func GenerateSafePrimes(nbits int, random io.Reader) (p, q *big.Int, err error) {
```
Don't we need to pass the `random io.Reader` to function below?

pdyraga replied:
I'll ☠️ this function.

pdyraga committed:
Fixed bit lenght assertions for safe prime generator tests

pdyraga committed:
Removed not needed code

pdyraga committed:
Improved safe prime generator documentation

pdyraga committed:
Removed deprecated GenerateSafePrimes function

The new function should be now called directly.

pdyraga committed:
Fixed typo in the documentation

pdyraga commented:
OK, this is ready for another chance!

pdyraga committed:
Deadlock fixed - result channel buffer increased

Although the result channel send operation and context cancel are
guarded by the same mutex and context is cancelled by the main thread
when it receives a result message from the channnel, there is absolutely
no guarantee that one goroutine can't write result to the channel after
the first goroutine did the same and before the main thread cancelled
the context. This is the following scenario:

- Worker #1 acquires mutex, writes to the result channel
- Worker #2 acquires mutex, writes to the result channel
- Main thread reads from the result channel, acquires mutex, cancels the
context

Since our result channel (and error channel) had buffer of size 1, the
Worker #2 was hanging indefinitely. Setting the buffer size to the
number of goroutines fixed the problem.

Alternative scenario was:

- Worker #1 acquires mutex, writes to the result channel
- Main thread acquires mutex, reads from the result channel, cancels the
context
- Worker #2 receives cancel notification so it completes without
yielding the result

That's why deadlock was happening only from time to time.

nkuba commented on thresholdkey_generator.go:
```diff
@@ -58,7 +59,7 @@ func GetThresholdKeyGenerator(nbits, TotalNumberOfDecryptionServers, Threshold i
 
 func (tkg *ThresholdKeyGenerator) generateSafePrimes() (*big.Int, *big.Int, error) {
-	return GenerateSafePrimes(tkg.nbits, tkg.Random)
+	return GenerateSafePrime(tkg.nbits, 4, 120*time.Second, tkg.Random)
```
What do you think about extracting the concurrency level value `4` to a constant on the file level?

pdyraga replied:
Yeah, I agree the `4` here is a kind of a magic number but extracting it to the file level gives an impression the entire threshold key generator is a multithread implementation while the only concurrent thing is the safe prime generator.

Let me propose the following instead:
```
func (tkg *ThresholdKeyGenerator) generateSafePrimes() (*big.Int, *big.Int, error) {
	concurrencyLevel := 4
	timeout := 120 * time.Second
	return GenerateSafePrime(tkg.nbits, concurrencyLevel, timeout, tkg.Random)
}
```

pdyraga committed:
Name parameters for the safe prime generator call

`4` and `120*time.Seconds` were just some magic numbers in the code.
We decided to move them to separate value declarations to make the
code more self-explanatory.

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+		cancel()
+		mutex.Unlock()
+		p, q, err = result.p, result.q, nil
+	case err := <-errChan:
```
This value of `err` never gets used because it's shadowed by the named return above (which is why I avoid named returns). Two options are:
1) Change this to `chanReturn` or another named error
2) Nuke the named returns, allocate them directly in the function, and return them explicitly. This will involve having explicit returns (which I prefer, though, depends on the style you're enforcing in this package).

pdyraga replied:
Yeah, good point. We'll need to use `defer waitGroup()` here to make sure it's called regardless which `return` path we go.

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+
+	bigMod := new(big.Int)
+
+	go func() {
```
You can put `defer waitGroup.Done()` here and remove the other two references. That way we know that we'll call `Done` regardless of what path we trigger!

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+				_, err := io.ReadFull(rand, bytes)
+				if err != nil {
+					errChan <- err
+					return
```
Nice! Very important to have the early returns ha

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+	defer close(primeChan)
+	defer close(errChan)
+
+	mutex := &sync.Mutex{}
```
I think you can get rid of this mutex. The context package is just constructors for the signal crossing API boundaries, and should be threadsafe in the way that you're using it. I'm not 100% on that, so I pulled your code, removed all references to `mutex`, and ran `go test -v -race -count=40` a few times just to be sure. All was fine.

pdyraga replied:
It sometimes fails without the mutex:
```
$ go test -count 100
panic: send on closed channel

goroutine 291 [running]:
github.com/keep-network/paillier.runGenPrimeRoutine.func1(0x1180e00, 0xc42001eb80, 0x11804c0, 0xc42008a2d0, 0xc4202523e4, 0x2, 0x2, 0xc4201c8420, 0x1, 0xc42040e980, ...)
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:272 +0x483
created by github.com/keep-network/paillier.runGenPrimeRoutine
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:179 +0x21c
exit status 2
FAIL    github.com/keep-network/paillier        1.082s
```

My theory is that worker thread is preempted after the `if ctx.Err() == nil ` check and someone cancels the context in a meantime. It results in closing the channel, and when thread has the control back, it can't `primeChan <- safePrime{p, q}`. 

However, this `mutex` was added before we got the `waitGroup.Wait()` in the main thread. My understanding how `waitGroup.Wait()`  and `defer close(primeChan)` works is that `close(primeChan)` should not be called before anything other in `GenerateSafePrime` including `waitGroup.Wait()`. So I don't really understand why is it possible that we closed the channel before we received `defer waitGroup.Done()` from all the workers and thus `waitGroup.Wait()` allowed to go further.

pdyraga replied:
My fault, instead of stacking changes I should try them one by one. The comment above is no longer valid. You are right, we can remove mutex and cancelled context check.

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+	}()
+}
+
+func isPocklingtonCriterionSatisfied(p *big.Int) bool {
```
A bit of documentation would help for this one.

pdyraga replied:
I added some minimal documentation here. The proof is not that easy so I did not rewrite it into the documentation: 
https://math.stackexchange.com/a/2697758

rargulati commented on .travis.yml:
```diff
@@ -1,6 +1,5 @@
 go:
-  - 1.3
-  - 1.4
+  - 1.7
```
Why not latest?

pdyraga replied:
`tip` is there too:
```
go:
  - 1.7
  - tip
```

In my understanding it lets us say "we support everything 1.7+"

rargulati replied:
Sounds good, thanks for the clarification.

rargulati commented on safe_prime_generator.go:
```diff
@@ -0,0 +1,301 @@
+					q.BitLen() == qBitLen {
+
+					mutex.Lock()
+					if ctx.Err() == nil {
```
Hm, interesting - so if we've closed the `primeChan` then a send to `primeChan` will panic. 
Fortunately, since you're checking the value of `ctx.Done` above, as soon as it receives a signal, it will unblock that channel and immediately return, causing execution to halt.

From running the race detector, I don't think we have a race between where the signal is sent to Done, but we're trying to pass in a value on `primeChan` as it hasn't panic'd on a 100 race runs. So we don't need this check, right?

pdyraga replied:
It sometimes panics without this check:
```
$ go test -count 100
panic: send on closed channel

goroutine 5 [running]:
github.com/keep-network/paillier.runGenPrimeRoutine.func1(0x1180ea0, 0xc4200ec100, 0x1180560, 0xc42008e2d0, 0xc4202320a8, 0x1, 0x1, 0xc4200a6180, 0x5, 0xc42000a120, ...)
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:272 +0x462
created by github.com/keep-network/paillier.runGenPrimeRoutine
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:179 +0x22c
exit status 2
FAIL    github.com/keep-network/paillier        0.067s
```

I think it's a similar situation as the one described earlier: https://github.com/keep-network/paillier/pull/8#discussion_r206121820

rargulati reviewed (changes requested):
Looks good! Some minor style changes and then other recommended changes for simplicity!

pdyraga committed:
Simplified concurrency mechanisms for goroutines

It does not longer make sense to have a mutex and cancelled context check
since we have a wait group and channels with buffers with the size equal
the number of workers.

Without the wait group we had to make sure no goroutine tries to write
to the closed channel. Now, channel is closed after all goroutines
completed (either by finding prime or cancelling the work), so there is
no chance someone tries to write to closed channel.

pdyraga committed:
Simplified how values are returned from the GenerateSafePrime

We can use `defer waitGroup.Wait()` to make sure it's called regardless
which path we go. This we, we can call `return ...` directly from all
select cases and we don't need named parameters.

pdyraga committed:
Added documentation to isPoclingtonCriterionSatisfied

rargulati commented on safe_prime_generator.go:
```diff
@@ -63,18 +63,19 @@ func GenerateSafePrime(
 	defer close(errChan)
-
-	waitGroup := &sync.WaitGroup{}
+	defer waitGroup.Wait()
```
My personal opinion is that we should stop waiting as soon as we fire off the last goroutine (which is the timeout goroutine). Is there a specific advantage we get from deferring this this?

rargulati replied:
So right here:
```
	ctx, cancel := context.WithCancel(context.Background())

	for i := 0; i < concurrencyLevel; i++ {
		waitGroup.Add(1)
		runGenPrimeRoutine(
			ctx, primeChan, errChan, waitGroup, random, bitLen,
		)
	}

	// Cancel after the specified timeout.
	go func() {
		time.Sleep(timeout)
		cancel()
	}()
	waitGroup.Wait()

	select {
	case result := <-primeChan:
...
...
```

Is what I have above still valid?

pdyraga replied:
Having the `waitGroup.Wait()` before the `select` means we'll need to wait for all goroutines to finish before we go to the `select`. And we don't want it :wink:

We want the first goroutine to win and cancel the rest, so the `waitGroup.Wait()` is just for graceful exiting. The ideal scenario is that the first goroutine finds a prime, write it to the result channel and then we `cancel()` the rest of them in the `select`. We do `waitGroup.Wait()` at the end to make sure they are all done.

rargulati replied:
Oh! Cool, got it! Yep, the `cancel()` makes sense (which then triggers) the `Wait()` to execute via the `defer`.

rargulati commented:
This is looking really good!

pdyraga commented:
So guys, let me do some final benchmark check after all those changes and if that proves everything is still fine, we can merge.

pdyraga commented:
# 512-bit
```
benchmark               old ns/op       new ns/op      delta
BenchmarkPrimeGen-8     2268969681      203780104      -91.02%
BenchmarkPrimeGen-8     11386138267     359325036      -96.84%
BenchmarkPrimeGen-8     2631451953      157727754      -94.01%
BenchmarkPrimeGen-8     1934699051      164243527      -91.51%
BenchmarkPrimeGen-8     10074871723     192326900      -98.09%
BenchmarkPrimeGen-8     5705400096      203981665      -96.42%
BenchmarkPrimeGen-8     5569562094      242117912      -95.65%
BenchmarkPrimeGen-8     1784282054      154775349      -91.33%
BenchmarkPrimeGen-8     1239902622      133648270      -89.22%
BenchmarkPrimeGen-8     2478345181      218190369      -91.20%
BenchmarkPrimeGen-8     3105669082      269501266      -91.32%
BenchmarkPrimeGen-8     2015272750      230842844      -88.55%
BenchmarkPrimeGen-8     2052936393      151189573      -92.64%
BenchmarkPrimeGen-8     2870424928      170988297      -94.04%
BenchmarkPrimeGen-8     6326295939      262912971      -95.84%
BenchmarkPrimeGen-8     6128458193      217260736      -96.45%
BenchmarkPrimeGen-8     1991237139      265375042      -86.67%
BenchmarkPrimeGen-8     4419054637      207724906      -95.30%
BenchmarkPrimeGen-8     1278097111      449248604      -64.85%
BenchmarkPrimeGen-8     1596391885      332082367      -79.20%
BenchmarkPrimeGen-8     2874842860      172948847      -93.98%
BenchmarkPrimeGen-8     2763618377      274258647      -90.08%
BenchmarkPrimeGen-8     1894357212      336403987      -82.24%
BenchmarkPrimeGen-8     2820765637      362911869      -87.13%
BenchmarkPrimeGen-8     5442844899      295884848      -94.56%
BenchmarkPrimeGen-8     2317770026      197804613      -91.47%
BenchmarkPrimeGen-8     13632503186     197842397      -98.55%
BenchmarkPrimeGen-8     1627820128      268519766      -83.50%
BenchmarkPrimeGen-8     1618299835      267506384      -83.47%
BenchmarkPrimeGen-8     1563058438      298371712      -80.91%
BenchmarkPrimeGen-8     6981079384      251532928      -96.40%
BenchmarkPrimeGen-8     8583457685      318334811      -96.29%
BenchmarkPrimeGen-8     2920009879      238789346      -91.82%
BenchmarkPrimeGen-8     2855556611      1624911923     -43.10%
BenchmarkPrimeGen-8     4192479325      242686849      -94.21%
BenchmarkPrimeGen-8     2124136761      251099733      -88.18%
BenchmarkPrimeGen-8     3354065514      211003123      -93.71%
BenchmarkPrimeGen-8     4793574227      234297990      -95.11%
BenchmarkPrimeGen-8     7926695778      218152628      -97.25%
BenchmarkPrimeGen-8     3783685714      255740341      -93.24%
BenchmarkPrimeGen-8     5769461617      283868378      -95.08%
BenchmarkPrimeGen-8     3899060159      1150253120     -70.50%
BenchmarkPrimeGen-8     8162547820      222932890      -97.27%
BenchmarkPrimeGen-8     1162491142      228113950      -80.38%
BenchmarkPrimeGen-8     1292424634      248727494      -80.75%
BenchmarkPrimeGen-8     2615398761      219047911      -91.62%
BenchmarkPrimeGen-8     3015784825      244386603      -91.90%
BenchmarkPrimeGen-8     2652024361      256228240      -90.34%
BenchmarkPrimeGen-8     2649694963      169500925      -93.60%
BenchmarkPrimeGen-8     1152981067      429531922      -62.75%

benchmark               old allocs     new allocs     delta
BenchmarkPrimeGen-8     362387         376490         +3.89%
BenchmarkPrimeGen-8     1769941        673830         -61.93%
BenchmarkPrimeGen-8     419616         264518         -36.96%
BenchmarkPrimeGen-8     311689         297350         -4.60%
BenchmarkPrimeGen-8     1653901        349166         -78.89%
BenchmarkPrimeGen-8     947327         349920         -63.06%
BenchmarkPrimeGen-8     917727         430910         -53.05%
BenchmarkPrimeGen-8     293204         276787         -5.60%
BenchmarkPrimeGen-8     204671         233339         +14.01%
BenchmarkPrimeGen-8     412135         385804         -6.39%
BenchmarkPrimeGen-8     512568         463953         -9.48%
BenchmarkPrimeGen-8     336663         380767         +13.10%
BenchmarkPrimeGen-8     339021         237222         -30.03%
BenchmarkPrimeGen-8     475326         274541         -42.24%
BenchmarkPrimeGen-8     1045261        394713         -62.24%
BenchmarkPrimeGen-8     1009762        321582         -68.15%
BenchmarkPrimeGen-8     328106         385440         +17.47%
BenchmarkPrimeGen-8     727409         269998         -62.88%
BenchmarkPrimeGen-8     202614         606475         +199.33%
BenchmarkPrimeGen-8     263268         435912         +65.58%
BenchmarkPrimeGen-8     470412         232267         -50.62%
BenchmarkPrimeGen-8     444871         372689         -16.23%
BenchmarkPrimeGen-8     325247         457537         +40.67%
BenchmarkPrimeGen-8     496687         495494         -0.24%
BenchmarkPrimeGen-8     932339         402878         -56.79%
BenchmarkPrimeGen-8     382644         265367         -30.65%
BenchmarkPrimeGen-8     2351353        268830         -88.57%
BenchmarkPrimeGen-8     281696         364910         +29.54%
BenchmarkPrimeGen-8     284146         361908         +27.37%
BenchmarkPrimeGen-8     271630         407238         +49.92%
BenchmarkPrimeGen-8     1205117        341747         -71.64%
BenchmarkPrimeGen-8     1510913        432134         -71.40%
BenchmarkPrimeGen-8     501141         323307         -35.49%
BenchmarkPrimeGen-8     497295         2162787        +334.91%
BenchmarkPrimeGen-8     744680         329554         -55.75%
BenchmarkPrimeGen-8     370638         337829         -8.85%
BenchmarkPrimeGen-8     601074         284034         -52.75%
BenchmarkPrimeGen-8     846588         313105         -63.02%
BenchmarkPrimeGen-8     1353487        303223         -77.60%
BenchmarkPrimeGen-8     638008         359257         -43.69%
BenchmarkPrimeGen-8     1012358        389266         -61.55%
BenchmarkPrimeGen-8     688078         1582064        +129.93%
BenchmarkPrimeGen-8     1424913        312111         -78.10%
BenchmarkPrimeGen-8     200156         319101         +59.43%
BenchmarkPrimeGen-8     224154         350082         +56.18%
BenchmarkPrimeGen-8     445941         297836         -33.21%
BenchmarkPrimeGen-8     535468         346711         -35.25%
BenchmarkPrimeGen-8     458854         360107         -21.52%
BenchmarkPrimeGen-8     459767         237852         -48.27%
BenchmarkPrimeGen-8     200718         602341         +200.09%

benchmark               old bytes     new bytes     delta
BenchmarkPrimeGen-8     82728176      38798343      -53.10%
BenchmarkPrimeGen-8     409631096     69508454      -83.03%
BenchmarkPrimeGen-8     96872624      27121609      -72.00%
BenchmarkPrimeGen-8     71888744      30673025      -57.33%
BenchmarkPrimeGen-8     378938424     35973569      -90.51%
BenchmarkPrimeGen-8     217741152     36060750      -83.44%
BenchmarkPrimeGen-8     212017904     44437788      -79.04%
BenchmarkPrimeGen-8     67207840      28515551      -57.57%
BenchmarkPrimeGen-8     47274152      24045820      -49.14%
BenchmarkPrimeGen-8     95460120      39786659      -58.32%
BenchmarkPrimeGen-8     118061851     47853263      -59.47%
BenchmarkPrimeGen-8     78798712      39294056      -50.13%
BenchmarkPrimeGen-8     78144912      24473498      -68.68%
BenchmarkPrimeGen-8     108414480     28313912      -73.88%
BenchmarkPrimeGen-8     241962472     40718964      -83.17%
BenchmarkPrimeGen-8     232679504     33150925      -85.75%
BenchmarkPrimeGen-8     74858416      39741979      -46.91%
BenchmarkPrimeGen-8     167936024     27831160      -83.43%
BenchmarkPrimeGen-8     46505936      62419490      +34.22%
BenchmarkPrimeGen-8     60196144      45135324      -25.02%
BenchmarkPrimeGen-8     108738512     23909754      -78.01%
BenchmarkPrimeGen-8     102350984     38434684      -62.45%
BenchmarkPrimeGen-8     74867877      47209044      -36.94%
BenchmarkPrimeGen-8     114624815     51063228      -55.45%
BenchmarkPrimeGen-8     218440752     41554600      -80.98%
BenchmarkPrimeGen-8     86905816      27387320      -68.49%
BenchmarkPrimeGen-8     541723888     27707700      -94.89%
BenchmarkPrimeGen-8     64802296      37643486      -41.91%
BenchmarkPrimeGen-8     64726184      37278543      -42.41%
BenchmarkPrimeGen-8     62328698      42082143      -32.48%
BenchmarkPrimeGen-8     279747264     35185698      -87.42%
BenchmarkPrimeGen-8     347825616     44506653      -87.20%
BenchmarkPrimeGen-8     114803440     33470376      -70.85%
BenchmarkPrimeGen-8     113669288     222001528     +95.30%
BenchmarkPrimeGen-8     172993448     34011901      -80.34%
BenchmarkPrimeGen-8     84498152      35026334      -58.55%
BenchmarkPrimeGen-8     138581872     29250029      -78.89%
BenchmarkPrimeGen-8     193495360     32231944      -83.34%
BenchmarkPrimeGen-8     311015920     31260818      -89.95%
BenchmarkPrimeGen-8     147068656     37023272      -74.83%
BenchmarkPrimeGen-8     234513488     40191003      -82.86%
BenchmarkPrimeGen-8     159871800     162752760     +1.80%
BenchmarkPrimeGen-8     329108176     32146308      -90.23%
BenchmarkPrimeGen-8     45833152      32874548      -28.27%
BenchmarkPrimeGen-8     51570752      36101255      -30.00%
BenchmarkPrimeGen-8     102344696     30658607      -70.04%
BenchmarkPrimeGen-8     123446618     35831864      -70.97%
BenchmarkPrimeGen-8     105297912     37145212      -64.72%
BenchmarkPrimeGen-8     106156028     24578604      -76.85%
BenchmarkPrimeGen-8     45458288      61916476      +36.21%
```

# 1024-bit
```
benchmark               old ns/op        new ns/op       delta
BenchmarkPrimeGen-8     36861702183      2028029411      -94.50%
BenchmarkPrimeGen-8     45959705486      4034424505      -91.22%
BenchmarkPrimeGen-8     32302799509      5857204466      -81.87%
BenchmarkPrimeGen-8     10559412608      1118208532      -89.41%
BenchmarkPrimeGen-8     26361757373      7741049970      -70.64%
BenchmarkPrimeGen-8     29501857227      10468787412     -64.51%
BenchmarkPrimeGen-8     231272508288     2951481919      -98.72%
BenchmarkPrimeGen-8     7657751078       3623114510      -52.69%
BenchmarkPrimeGen-8     51558997401      9734591947      -81.12%
BenchmarkPrimeGen-8     16745757300      3662106162      -78.13%
BenchmarkPrimeGen-8     28182579779      4186072505      -85.15%
BenchmarkPrimeGen-8     19612730893      8350284991      -57.42%
BenchmarkPrimeGen-8     24840617742      3549950465      -85.71%
BenchmarkPrimeGen-8     2466813963       1229082030      -50.18%
BenchmarkPrimeGen-8     8494698198       3489757718      -58.92%
BenchmarkPrimeGen-8     24406594314      6860867774      -71.89%
BenchmarkPrimeGen-8     111450059549     4412554670      -96.04%
BenchmarkPrimeGen-8     9711480641       1544705780      -84.09%
BenchmarkPrimeGen-8     81936301479      2011204182      -97.55%
BenchmarkPrimeGen-8     2767979308       5385645947      +94.57%
BenchmarkPrimeGen-8     101696370631     4599258095      -95.48%
BenchmarkPrimeGen-8     22492110644      8224942742      -63.43%
BenchmarkPrimeGen-8     84566338764      8518647759      -89.93%
BenchmarkPrimeGen-8     17797979625      1120075567      -93.71%
BenchmarkPrimeGen-8     28036722072      2624979440      -90.64%
BenchmarkPrimeGen-8     26167954580      5402545689      -79.35%
BenchmarkPrimeGen-8     16405363943      4549220558      -72.27%
BenchmarkPrimeGen-8     4533869408       6277611905      +38.46%
BenchmarkPrimeGen-8     28292850899      3154577090      -88.85%
BenchmarkPrimeGen-8     58376102081      5058555291      -91.33%
BenchmarkPrimeGen-8     137515732119     8872658389      -93.55%
BenchmarkPrimeGen-8     5343562821       6414989815      +20.05%
BenchmarkPrimeGen-8     4214167930       1631263654      -61.29%
BenchmarkPrimeGen-8     57539380856      15259652544     -73.48%
BenchmarkPrimeGen-8     56463189239      6586140711      -88.34%
BenchmarkPrimeGen-8     31148876342      1541031706      -95.05%
BenchmarkPrimeGen-8     35924957141      9501304286      -73.55%
BenchmarkPrimeGen-8     37037673636      7761418403      -79.04%
BenchmarkPrimeGen-8     123071912422     6229731376      -94.94%
BenchmarkPrimeGen-8     18735582572      3162629339      -83.12%
BenchmarkPrimeGen-8     14950810812      1312778249      -91.22%
BenchmarkPrimeGen-8     45614828915      4644812942      -89.82%
BenchmarkPrimeGen-8     38343598558      2865488956      -92.53%
BenchmarkPrimeGen-8     37089393514      2191057855      -94.09%
BenchmarkPrimeGen-8     21408341771      3976707810      -81.42%
BenchmarkPrimeGen-8     82520362962      3960534907      -95.20%
BenchmarkPrimeGen-8     71017278949      3767476624      -94.69%
BenchmarkPrimeGen-8     34683384212      3285594016      -90.53%
BenchmarkPrimeGen-8     176688401313     10007708014     -94.34%
BenchmarkPrimeGen-8     27576279959      1812340656      -93.43%

benchmark               old allocs     new allocs     delta
BenchmarkPrimeGen-8     1296874        915655         -29.40%
BenchmarkPrimeGen-8     1629910        1813474        +11.26%
BenchmarkPrimeGen-8     1145560        2578585        +125.09%
BenchmarkPrimeGen-8     375806         442742         +17.81%
BenchmarkPrimeGen-8     936175         3080384        +229.04%
BenchmarkPrimeGen-8     1032492        4053990        +292.64%
BenchmarkPrimeGen-8     7860859        1128439        -85.64%
BenchmarkPrimeGen-8     278511         1335743        +379.60%
BenchmarkPrimeGen-8     1772920        3465450        +95.47%
BenchmarkPrimeGen-8     599647         1200613        +100.22%
BenchmarkPrimeGen-8     974381         1303572        +33.78%
BenchmarkPrimeGen-8     684946         2450948        +257.83%
BenchmarkPrimeGen-8     895839         986127         +10.08%
BenchmarkPrimeGen-8     84866          348326         +310.44%
BenchmarkPrimeGen-8     292856         980527         +234.82%
BenchmarkPrimeGen-8     865260         1752853        +102.58%
BenchmarkPrimeGen-8     3960309        1258875        -68.21%
BenchmarkPrimeGen-8     348814         439299         +25.94%
BenchmarkPrimeGen-8     2872539        521117         -81.86%
BenchmarkPrimeGen-8     96468          1559819        +1516.93%
BenchmarkPrimeGen-8     3535318        1460611        -58.69%
BenchmarkPrimeGen-8     774395         2542689        +228.35%
BenchmarkPrimeGen-8     2807298        2581099        -8.06%
BenchmarkPrimeGen-8     551587         352002         -36.18%
BenchmarkPrimeGen-8     909261         818918         -9.94%
BenchmarkPrimeGen-8     859315         1803050        +109.82%
BenchmarkPrimeGen-8     549266         1447562        +163.54%
BenchmarkPrimeGen-8     155427         1929606        +1141.49%
BenchmarkPrimeGen-8     976922         980262         +0.34%
BenchmarkPrimeGen-8     2063480        1695845        -17.82%
BenchmarkPrimeGen-8     4841641        2835719        -41.43%
BenchmarkPrimeGen-8     183380         2059248        +1022.94%
BenchmarkPrimeGen-8     147059         535376         +264.06%
BenchmarkPrimeGen-8     2041437        5092646        +149.46%
BenchmarkPrimeGen-8     1975536        2219626        +12.36%
BenchmarkPrimeGen-8     1074994        499865         -53.50%
BenchmarkPrimeGen-8     1287402        3238482        +151.55%
BenchmarkPrimeGen-8     1341691        2658899        +98.18%
BenchmarkPrimeGen-8     4409132        2130144        -51.69%
BenchmarkPrimeGen-8     658803         1005215        +52.58%
BenchmarkPrimeGen-8     516970         432836         -16.27%
BenchmarkPrimeGen-8     1628919        1550246        -4.83%
BenchmarkPrimeGen-8     1358365        981565         -27.74%
BenchmarkPrimeGen-8     1300251        787778         -39.41%
BenchmarkPrimeGen-8     773588         1350405        +74.56%
BenchmarkPrimeGen-8     2933710        1349526        -54.00%
BenchmarkPrimeGen-8     2551294        1288342        -49.50%
BenchmarkPrimeGen-8     1139499        1145489        +0.53%
BenchmarkPrimeGen-8     6096582        3471008        -43.07%
BenchmarkPrimeGen-8     974002         609313         -37.44%

benchmark               old bytes      new bytes     delta
BenchmarkPrimeGen-8     410531984      120005968     -70.77%
BenchmarkPrimeGen-8     516871536      237446256     -54.06%
BenchmarkPrimeGen-8     363488520      334296984     -8.03%
BenchmarkPrimeGen-8     118047720      57559776      -51.24%
BenchmarkPrimeGen-8     297614840      404117232     +35.79%
BenchmarkPrimeGen-8     328461424      530639784     +61.55%
BenchmarkPrimeGen-8     2498362152     146655272     -94.13%
BenchmarkPrimeGen-8     88981416       174243768     +95.82%
BenchmarkPrimeGen-8     562466976      452768648     -19.50%
BenchmarkPrimeGen-8     190238184      156798365     -17.58%
BenchmarkPrimeGen-8     309543552      172290192     -44.34%
BenchmarkPrimeGen-8     217364304      321025276     +47.69%
BenchmarkPrimeGen-8     284504936      129471416     -54.49%
BenchmarkPrimeGen-8     26543528       45600056      +71.79%
BenchmarkPrimeGen-8     92278832       128537784     +39.29%
BenchmarkPrimeGen-8     273084064      228649920     -16.27%
BenchmarkPrimeGen-8     1260919112     163888968     -87.00%
BenchmarkPrimeGen-8     110596312      58035824      -47.52%
BenchmarkPrimeGen-8     908471736      67798256      -92.54%
BenchmarkPrimeGen-8     31164144       202952584     +551.24%
BenchmarkPrimeGen-8     1126090384     191120288     -83.03%
BenchmarkPrimeGen-8     245900104      332266464     +35.12%
BenchmarkPrimeGen-8     892851912      337539136     -62.20%
BenchmarkPrimeGen-8     176047536      45909716      -73.92%
BenchmarkPrimeGen-8     289382520      107164232     -62.97%
BenchmarkPrimeGen-8     273224016      236085232     -13.59%
BenchmarkPrimeGen-8     173397688      190356416     +9.78%
BenchmarkPrimeGen-8     48816216       251743656     +415.70%
BenchmarkPrimeGen-8     309511600      127547596     -58.79%
BenchmarkPrimeGen-8     654125952      221186192     -66.19%
BenchmarkPrimeGen-8     1536756272     371946200     -75.80%
BenchmarkPrimeGen-8     57621800       271916640     +371.90%
BenchmarkPrimeGen-8     46709232       69726392      +49.28%
BenchmarkPrimeGen-8     646374992      665026640     +2.89%
BenchmarkPrimeGen-8     624530576      289691416     -53.61%
BenchmarkPrimeGen-8     342230304      65982736      -80.72%
BenchmarkPrimeGen-8     411465472      423620944     +2.95%
BenchmarkPrimeGen-8     427305360      348680728     -18.40%
BenchmarkPrimeGen-8     1403440376     278391536     -80.16%
BenchmarkPrimeGen-8     208987312      132752456     -36.48%
BenchmarkPrimeGen-8     163475552      56851360      -65.22%
BenchmarkPrimeGen-8     516464392      202416184     -60.81%
BenchmarkPrimeGen-8     432335688      128268376     -70.33%
BenchmarkPrimeGen-8     414238240      102396720     -75.28%
BenchmarkPrimeGen-8     246773984      177088104     -28.24%
BenchmarkPrimeGen-8     927995536      176423193     -80.99%
BenchmarkPrimeGen-8     813041256      169167776     -79.19%
BenchmarkPrimeGen-8     360886104      149682264     -58.52%
BenchmarkPrimeGen-8     1934911896     453022168     -76.59%
BenchmarkPrimeGen-8     309722576      80011776      -74.17%
```

pdyraga commented:
We are still good performance-wise :tada: @rargulati or @nkuba would you like to do the honors?
